src/gsy_e/models/market/forward.py

```diff
@@ -0,0 +1,138 @@
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+from abc import abstractmethod
+from calendar import monthrange
+from typing import Optional, TYPE_CHECKING
+
+from gsy_framework.constants_limits import ConstSettings
+from pendulum import DateTime, duration
+
+from gsy_e.gsy_e_core.blockchain_interface import NonBlockchainInterface
+from gsy_e.models.market import GridFee
+from gsy_e.models.market.future import FutureMarkets
+
+if TYPE_CHECKING:
+    from gsy_e.models.config import SimulationConfig
+    from gsy_e.models.area.event_dispatcher import AreaDispatcher
+
+
+class ForwardMarketBase(FutureMarkets):
+    """Base class for forward markets"""
+
+    def __init__(self, bc: Optional[NonBlockchainInterface] = None,
+                 notification_listener: Optional["AreaDispatcher"] = None,
+                 readonly: bool = False,
+                 grid_fee_type: int = ConstSettings.MASettings.GRID_FEE_TYPE,
+                 grid_fees: Optional[GridFee] = None,
+                 name: Optional[str] = None) -> None:
+        # pylint: disable=too-many-arguments
+        if not ConstSettings.ForwardMarketSettings.ENABLE_FORWARD_MARKETS:
+            return
+
+        super().__init__(bc=bc, notification_listener=notification_listener,
+                         readonly=readonly, grid_fee_type=grid_fee_type,
+                         grid_fees=grid_fees, name=name)
+
+    @staticmethod
+    @abstractmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        """Return time when the market block starts."""
+
+    @staticmethod
+    @abstractmethod
+    def _get_end_time(current_time: DateTime) -> DateTime:
+        """Return time when the market block ends."""
+
+    @staticmethod
+    @abstractmethod
+    def _get_market_slot_duration(current_time: DateTime, _config) -> duration:
+        """Return duration of market slots inside the market block."""
+
+    def create_future_market_slots(self, current_market_time_slot: DateTime,
+                                   config: "SimulationConfig") -> None:
+        if not ConstSettings.ForwardMarketSettings.ENABLE_FORWARD_MARKETS:
+            return
+        self._create_future_market_slots(self._get_start_time(current_market_time_slot),
+                                         self._get_end_time(current_market_time_slot), config)
+
+
+class DayForwardMarket(ForwardMarketBase):
+    """Day forward market implementation"""
+
+    @staticmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        return current_time.set(hour=0, minute=0).add(days=1)
```

Comment:
Change the order of `.set` and `.add` as discussed below in `YearForwardMarket`. Thanks

Reply:
@spyrostz How did we decide the `.add(days=1)` here? 

AFAIU, if the daily market is on the `2022-07-28` (all day), it should open on the  `2022-07-21T00:00` (one week in advance) and it should close on the `2022-07-27T23:00:00` (one hour before the first market slot). But how did we decide to create these future market slots exactly 1 day in advance?

Reply:
@fievelk I think that we have more or less discussed the same thing in other comments too, but IMO this implementation is correct. All day-ahead (hourly) markets for one day (2022-07-28 in your example) should open 7 days in advance and close one hour before the day. If we did not have the `add(days=1)` here, each hourly market for the same day would start at a different time and not at exactly 00:00 as the requirements dictate.

Reply:
@spyrostz I'm not sure we're on the same topic here. My question would basically be: why do we use `.add(days=1)` and not (for example) `.add(days=5)` (just random)? This should just decide how much in advance we create the market slots, but AFAIU it doesn't affect the duration of the open market. So I just wanted to double check: if the behaviour is the same, this `.add(days=1)` could be anything >=1, right?

Reply:
@fievelk now I got your question, in theory you are right, we could even have `.add(days=12243)` and it would still work as expected.